File: android/src/main/kotlin/kr/yhs/flutter_kakaomaps/controller/overlay/handler/LodLabelControllerHandler.kt
package kr.yhs.flutter_kakaomaps.controller.overlay.handler

import com.kakao.vectormap.label.LabelLayerOptions
import com.kakao.vectormap.label.LabelManager
import com.kakao.vectormap.label.LabelOptions
import com.kakao.vectormap.label.LodLabel
import com.kakao.vectormap.label.LodLabelLayer
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kr.yhs.flutter_kakaomaps.converter.LabelTypeConverter.asLabelLayerOptions
import kr.yhs.flutter_kakaomaps.converter.LabelTypeConverter.asLabelOptions
import kr.yhs.flutter_kakaomaps.converter.PrimitiveTypeConverter.asBoolean
import kr.yhs.flutter_kakaomaps.converter.PrimitiveTypeConverter.asLong
import kr.yhs.flutter_kakaomaps.converter.PrimitiveTypeConverter.asMap
import kr.yhs.flutter_kakaomaps.converter.PrimitiveTypeConverter.asString

interface LodLabelControllerHandler {
    val labelManager: LabelManager?

    fun lodLabelHandle(call: MethodCall, result: MethodChannel.Result) {
        val arguments = call.arguments!!.asMap<Any?>()
        if (labelManager == null) {
            throw NullPointerException("LabelManager is null.")
        }

        val layer =
                arguments["layerId"]?.asString()?.let<String, LodLabelLayer> {
                    labelManager!!.getLodLayer(it)
                }
        val poi = layer?.run { arguments["poiId"]?.asString()?.let(layer::getLabel) }

        when (call.method) {
            "createLodLabelLayer" -> {
                createLodLabelLayer(arguments.asLabelLayerOptions(), result::success)
            }
            "removeLodLabelLayer" -> {
                removeLodLabelLayer(layer!!, result::success)
            }
            "addLodPoi" -> {
                val poiOption = arguments["poi"]!!.asLabelOptions(labelManager!!)
                addLodPoi(layer!!, poiOption, result::success)
            }
            "removeLodPoi" -> removeLodPoi(layer!!, poi!!, result::success)
            // poi Handler
            "changePoiVisible" -> {
                val visible = arguments["visible"]?.asBoolean()!!
                changeLodPoiVisible(poi!!, visible, result::success)
            }
            "changePoiStyle" -> {
                val styleId = arguments["styleId"]?.asString()!!
                changeLodPoiStyle(poi!!, styleId, result::success)
            }
            "changePoiText" -> {
                val text = arguments["text"]?.asString()!!
                changeLodPoiStyle(poi!!, text, result::success)
            }
            "rankPoi" -> {
                val rank = arguments["x"]?.asLong()!!
                rankLodPoi(poi!!, rank, result::success)
            }
            else -> result.notImplemented()
        }
    }

    fun createLodLabelLayer(options: LabelLayerOptions, onSuccess: (Any?) -> Unit)

    fun removeLodLabelLayer(layer: LodLabelLayer, onSuccess: (Any?) -> Unit)

    fun addLodPoi(layer: LodLabelLayer, poi: LabelOptions, onSuccess: (String) -> Unit)

    fun removeLodPoi(layer: LodLabelLayer, poi: LodLabel, onSuccess: (Any?) -> Unit)

    // Lod-Poi Controller
    fun changeLodPoiVisible(poi: LodLabel, visible: Boolean, onSuccess: (Any?) -> Unit)

    fun changeLodPoiStyle(poi: LodLabel, styleId: String, onSuccess: (Any?) -> Unit)

    fun changeLodPoiText(poi: LodLabel, text: String, onSuccess: (Any?) -> Unit)

    fun rankLodPoi(poi: LodLabel, rank: Long, onSuccess: (Any?) -> Unit)
}

File: ios/Classes/controller/overlay/handler/LodLabelControllerHandler.swift
import KakaoMapsSDK
import Flutter

internal protocol LodLabelControllerHandler {
    var labelManager: LabelManager { get };

    func createLodLabelLayer(option: LodLabelLayerOptions, onSuccess: (Any?) -> Void)

    func removeLodLabelLayer(layerId: String, onSuccess: (Any?) -> Void)

    func addLodPoi(layer: LodLabelLayer, poi: PoiOptions, position: MapPoint, visible: Bool, onSuccess: @escaping (String) -> Void)

    func removeLodPoi(layer: LodLabelLayer, poiId: String, onSuccess: (Any?) -> Void)
    
    func changeLodPoiVisible(poi: LodPoi, visible: Bool, autoMove: Bool, onSuccess: (Any?) -> Void)

    func changeLodPoiStyle(poi: LodPoi, styleId: String, transition: Bool, onSuccess: (Any?) -> Void)

    func changeLodPoiText(poi: LodPoi, styleId: String, text: String, transition: Bool, onSuccess: (Any?) -> Void)

    func rankLodPoi(poi: LodPoi, rank: Int, onSuccess: (Any?) -> Void)
}

internal extension LodLabelControllerHandler {
    func lodLabelHandle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = castSafty(call.arguments, caster: asDict)
        let layerId: String? = castSafty(arguments?["layerId"], caster: asString)
        let layer: LodLabelLayer? = layerId.flatMap { key in
            labelManager.getLodLabelLayer(layerID: key)
        }

        let poiId = castSafty(arguments?["poiId"], caster: asString)
        let poi: LodPoi? = poiId.flatMap { key in
            layer!.getLodPoi(poiID: key)
        }

        switch call.method {
        case "createLodLabelLayer": createLodLabelLayer(option: LodLabelLayerOptions(payload: arguments!), onSuccess: result)
        case "removeLodLabelLayer": removeLodLabelLayer(layerId: layerId!, onSuccess: result)
        case "addLodPoi":
            let poiArgument = asDict(arguments!["poi"]!)
            let poiOption = PoiOptions(payload: poiArgument)
            let position = MapPoint(payload: poiArgument)
            let visible = asBool(arguments!["visible"] ?? true)
            addLodPoi(layer: layer!, poi: PoiOptions(payload: poiArgument), position: position, visible: visible, onSuccess: result)
        case "removeLodPoi": removeLodPoi(layer: layer!, poiId: poiId!, onSuccess: result)
        case "changePoiVisible":
            let visible = asBool(arguments!["visible"]!)
            let autoMove = castSafty(arguments!["autoMove"], caster: asBool) ?? false
            changeLodPoiVisible(poi: poi!, visible: visible, autoMove: autoMove, onSuccess: result)
        case "changePoiStyle":
            let styleId = asString(arguments!["styleId"]!)
            let transition = asBool(arguments!["transition"] ?? false)
            changeLodPoiStyle(poi: poi!, styleId: styleId, transition: transition, onSuccess: result)
        case "changePoiText":
            let text = asString(arguments!["text"]!)
            let transition = asBool(arguments!["transition"] ?? false)
            let styleId = asString(arguments!["styleId"]!)
            changeLodPoiText(poi: poi!, styleId: styleId, text: text, transition: transition, onSuccess: result)
        case "rankPoi":
            let rank = asInt(arguments!["rank"]!)
            rankLodPoi(poi: poi!, rank: rank, onSuccess: result)
        default: result(FlutterMethodNotImplemented)
        }
    }
}
